File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/java/com/divideconquer/algorithms/StrassenMatrix.java
package com.divideconquer.algorithms;

public final class StrassenMatrix {
    private StrassenMatrix() {
    }

    public static int[][] multiply(int[][] a, int[][] b) {
        validate(a, b);
        int n = a.length;
        int paddedSize = nextPowerOfTwo(n);
        if (paddedSize != n) {
            int[][] result = multiplyPowerOfTwo(pad(a, paddedSize), pad(b, paddedSize));
            return trim(result, n);
        }
        return multiplyPowerOfTwo(a, b);
    }

    private static int[][] multiplyPowerOfTwo(int[][] a, int[][] b) {
        int n = a.length;
        if (n == 1) {
            return new int[][]{{a[0][0] * b[0][0]}};
        }

        int half = n / 2;
        int[][] a11 = slice(a, 0, 0, half);
        int[][] a12 = slice(a, 0, half, half);
        int[][] a21 = slice(a, half, 0, half);
        int[][] a22 = slice(a, half, half, half);
        int[][] b11 = slice(b, 0, 0, half);
        int[][] b12 = slice(b, 0, half, half);
        int[][] b21 = slice(b, half, 0, half);
        int[][] b22 = slice(b, half, half, half);

        int[][] p1 = multiplyPowerOfTwo(add(a11, a22), add(b11, b22));
        int[][] p2 = multiplyPowerOfTwo(add(a21, a22), b11);
        int[][] p3 = multiplyPowerOfTwo(a11, subtract(b12, b22));
        int[][] p4 = multiplyPowerOfTwo(a22, subtract(b21, b11));
        int[][] p5 = multiplyPowerOfTwo(add(a11, a12), b22);
        int[][] p6 = multiplyPowerOfTwo(subtract(a21, a11), add(b11, b12));
        int[][] p7 = multiplyPowerOfTwo(subtract(a12, a22), add(b21, b22));

        int[][] c11 = add(subtract(add(p1, p4), p5), p7);
        int[][] c12 = add(p3, p5);
        int[][] c21 = add(p2, p4);
        int[][] c22 = add(subtract(add(p1, p3), p2), p6);

        return join(c11, c12, c21, c22);
    }

    private static void validate(int[][] a, int[][] b) {
        if (a == null || b == null || a.length == 0 || a.length != b.length) {
            throw new IllegalArgumentException("matrices must be non-empty and have the same square size");
        }
        int n = a.length;
        for (int i = 0; i < n; i++) {
            if (a[i] == null || b[i] == null || a[i].length != n || b[i].length != n) {
                throw new IllegalArgumentException("matrices must be square");
            }
        }
    }

    private static int nextPowerOfTwo(int value) {
        int result = 1;
        while (result < value) {
            result *= 2;
        }
        return result;
    }

    private static int[][] pad(int[][] matrix, int size) {
        int[][] result = new int[size][size];
        for (int row = 0; row < matrix.length; row++) {
            System.arraycopy(matrix[row], 0, result[row], 0, matrix.length);
        }
        return result;
    }

    private static int[][] trim(int[][] matrix, int size) {
        int[][] result = new int[size][size];
        for (int row = 0; row < size; row++) {
            System.arraycopy(matrix[row], 0, result[row], 0, size);
        }
        return result;
    }

    private static int[][] add(int[][] a, int[][] b) {
        int n = a.length;
        int[][] result = new int[n][n];
        for (int row = 0; row < n; row++) {
            for (int col = 0; col < n; col++) {
                result[row][col] = a[row][col] + b[row][col];
            }
        }
        return result;
    }

    private static int[][] subtract(int[][] a, int[][] b) {
        int n = a.length;
        int[][] result = new int[n][n];
        for (int row = 0; row < n; row++) {
            for (int col = 0; col < n; col++) {
                result[row][col] = a[row][col] - b[row][col];
            }
        }
        return result;
    }

    private static int[][] slice(int[][] matrix, int startRow, int startCol, int size) {
        int[][] result = new int[size][size];
        for (int row = 0; row < size; row++) {
            System.arraycopy(matrix[startRow + row], startCol, result[row], 0, size);
        }
        return result;
    }

    private static int[][] join(int[][] c11, int[][] c12, int[][] c21, int[][] c22) {
        int half = c11.length;
        int[][] result = new int[half * 2][half * 2];
        for (int row = 0; row < half; row++) {
            System.arraycopy(c11[row], 0, result[row], 0, half);
            System.arraycopy(c12[row], 0, result[row], half, half);
            System.arraycopy(c21[row], 0, result[half + row], 0, half);
            System.arraycopy(c22[row], 0, result[half + row], half, half);
        }
        return result;
    }
}

File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/python/divide_conquer/algorithms/strassen_matrix.py
"""Strassen matrix multiplication for square matrices."""

Matrix = list[list[int]]


def multiply(a: Matrix, b: Matrix) -> Matrix:
    """Multiply two square matrices, padding non-power-of-two sizes internally."""
    _validate(a, b)
    n = len(a)
    padded_size = _next_power_of_two(n)
    if padded_size != n:
        return _trim(_multiply_power_of_two(_pad(a, padded_size), _pad(b, padded_size)), n)
    return _multiply_power_of_two(a, b)


def _multiply_power_of_two(a: Matrix, b: Matrix) -> Matrix:
    n = len(a)
    if n == 1:
        return [[a[0][0] * b[0][0]]]

    half = n // 2
    a11 = _slice(a, 0, 0, half)
    a12 = _slice(a, 0, half, half)
    a21 = _slice(a, half, 0, half)
    a22 = _slice(a, half, half, half)
    b11 = _slice(b, 0, 0, half)
    b12 = _slice(b, 0, half, half)
    b21 = _slice(b, half, 0, half)
    b22 = _slice(b, half, half, half)

    p1 = _multiply_power_of_two(_add(a11, a22), _add(b11, b22))
    p2 = _multiply_power_of_two(_add(a21, a22), b11)
    p3 = _multiply_power_of_two(a11, _subtract(b12, b22))
    p4 = _multiply_power_of_two(a22, _subtract(b21, b11))
    p5 = _multiply_power_of_two(_add(a11, a12), b22)
    p6 = _multiply_power_of_two(_subtract(a21, a11), _add(b11, b12))
    p7 = _multiply_power_of_two(_subtract(a12, a22), _add(b21, b22))

    c11 = _add(_subtract(_add(p1, p4), p5), p7)
    c12 = _add(p3, p5)
    c21 = _add(p2, p4)
    c22 = _add(_subtract(_add(p1, p3), p2), p6)
    return _join(c11, c12, c21, c22)


def _validate(a: Matrix, b: Matrix) -> None:
    if a is None or b is None:
        raise ValueError("matrices must not be None")
    if not a or len(a) != len(b):
        raise ValueError("matrices must be non-empty and have the same square size")
    n = len(a)
    if any(row is None or len(row) != n for row in a) or any(row is None or len(row) != n for row in b):
        raise ValueError("matrices must be square")


def _next_power_of_two(value: int) -> int:
    result = 1
    while result < value:
        result *= 2
    return result


def _pad(matrix: Matrix, size: int) -> Matrix:
    result = [[0 for _ in range(size)] for _ in range(size)]
    for row_index, row in enumerate(matrix):
        result[row_index][: len(matrix)] = row
    return result


def _trim(matrix: Matrix, size: int) -> Matrix:
    return [row[:size] for row in matrix[:size]]


def _add(a: Matrix, b: Matrix) -> Matrix:
    return [[a[row][col] + b[row][col] for col in range(len(a))] for row in range(len(a))]


def _subtract(a: Matrix, b: Matrix) -> Matrix:
    return [[a[row][col] - b[row][col] for col in range(len(a))] for row in range(len(a))]


def _slice(matrix: Matrix, row: int, col: int, size: int) -> Matrix:
    return [matrix[index][col : col + size] for index in range(row, row + size)]


def _join(c11: Matrix, c12: Matrix, c21: Matrix, c22: Matrix) -> Matrix:
    top = [left + right for left, right in zip(c11, c12, strict=True)]
    bottom = [left + right for left, right in zip(c21, c22, strict=True)]
    return top + bottom
